add content filter to filter the article by keyword expression

Jinwei1987 opened the pull request:
add content filter to filter the article by keyword expression
it will be able to support expression like 'key1 | key2 | key3' or more complex one like '(key1 & key2) | (key3 & key4)'

fhamborg commented on newsplease/pipeline/pipelines.py:
```diff
@@ -570,6 +571,147 @@ def process_item(self, item, spider):
 
+class ContentFilter(object):
+
+    class BiOp(object):
```
Could you please use a better understandable class name?

fhamborg replied:
and briefly state what this class is doing

Jinwei1987 replied:
have added the comments

fhamborg commented on newsplease/config/config.cfg:
```diff
@@ -263,6 +263,13 @@ end_date = '2999-12-31 00:00:00'
+[ContentFilter]
+
+filter_predicates = {'title': 'cloud computing',
+                     'text': 'google | amazon'}
```
Can you add a brief comment to explain how this is to be used? E.g., that you can use the term filter on either the title or the text field.

fhamborg replied:
Also, can one use pure strings or also regex?

Jinwei1987 replied:
Pure string without &, | and ( is supported. 
regex is not supported. I feel logic expression is easier to use comparing with regex. with nested logic expression, there will be rare case where regex is needed

fhamborg commented on newsplease/config/config.cfg:
```diff
@@ -263,6 +263,13 @@ end_date = '2999-12-31 00:00:00'
+
+filter_predicates = {'title': 'cloud computing',
+                     'text': 'google | amazon'}
+pass_condition = title or text
```
Please specify as a comment any other values that one can put here besides title or text.

fhamborg commented:
Hi! Thank you for the PR! I've added a few comments in order to clear up a few minor things. I really appreciate your PR and support :-)

Jinwei1987 commented:
> Hi! Thank you for the PR! I've added a few comments in order to clear up a few minor things. I really appreciate your PR and support :-)

I have added the description of the class and moments to explain the use cases

fhamborg commented:
We finally had a news-please core developer meeting so that I can give an update on this. While we still want to add the filter functionality, we think that using regex solely is the most suitable approach. Compared to the approach in this PR, regex not only supports basic logic functionality but much more in terms of string and pattern matching. Thus, we cannot merge the PR in this form into the master branch. Still, thank you for the PR! If you are considering removing the biop functionality and use regex instead, please just let me know. Thank you!
